src/os/FileStore.cc

```diff
@@ -1830,13 +1840,65 @@ void FileStore::queue_op(OpSequencer *osr, Op *o)
   dout(5) << "queue_op " << o << " seq " << o->op
 	  << " " << *osr
 	  << " " << o->bytes << " bytes"
-	  << "   (queue has " << throttle_ops.get_current() << " ops and " << throttle_bytes.get_current() << " bytes)"
+	  << "   (queue has " << throttle_ops.get_current() << " ops and " 
+          << throttle_bytes.get_current() << " bytes)"
 	  << dendl;
   op_wq.queue(osr);
+
+  if (g_conf->journal_dynamic_throttle) {
+  /* Trying to throttle the journal commit based on the number of
+     outstanding bytes to be applied. This will restrict the amount of
+     memory growth in the write path. The mode of throttle is to introduce
+     an incremental delay based on the % increase above an outstanding byte
+     threshold.
+   */
+    uint64_t outstanding_bytes = m_filestore_current_queue_bytes.read(); 
+    bool induce_delay = false;
+    uint32_t percentage_more = 0;
+    
+    if (outstanding_bytes > m_filestore_queue_max_bytes) {
+      percentage_more = (uint32_t)((100* (outstanding_bytes -
+                      m_filestore_queue_max_bytes))/m_filestore_queue_max_bytes);
+      if (percentage_more > 100) {
+        percentage_more = 100;
+      }
+
+      induce_delay = true;
+    }
+    
+    if (induce_delay) {
+
+      double delay_time;
+      dyn_throttle_filestore.calc_wait_time(100 - percentage_more, 
+                                                  delay_time, true);
+      if (0 != delay_time) {
+        struct timespec waittime;
+        waittime.tv_sec = 0;
+        waittime.tv_nsec = delay_time * 1000 * 1000 * 1000 ;
+        nanosleep(&waittime, NULL);
+        dout(5) << " Filestore on BaseDir = " << basedir << " and in journal on "
+            << journalpath << " is slow on applying transaction."
+            << " Throttling journal commits.Filestore op_wq depth = "
+            << op_queue.size() << " Outstanding bytes to be applied = "
+            << outstanding_bytes << " seq = " << o->op
+            << " Max op threshold = " << m_filestore_queue_max_ops
+            << " Max byte threshold = " << m_filestore_queue_max_bytes
+            << " Transaction size = " << o->bytes
+            << " Percentage exceeded threshold = " << percentage_more
+            << " Applying delay = " << delay_time << " in the journal commit path."
+            << dendl;
```

Comment:
Yes, I wanted to go away from the model that will do nothing unless a threshold value is crossed and then pausing everything to bring the value down below the threshold.
New model is to induce small delay based on the following scenarios:

1. Based on journal % emptiness

2. Will induce an order of magnitude more delay if % emptiness goes below a threshold

3. If the outstanding Tx bytes in the filestore level goes above a threshold (filestore_queue_max_bytes), based on the % above it is again inducing a delay.

This is benefiting performance significantly specially if the storage is fast and in case of journal is on different storage device (tested with NVRAM as journal).
But, the side effect is, if we don't tune this properly and make journal go unleash , there will be  a memory growth. This is one of the reason, I am not proposing this as a default as user should be knowing what they are doing.
The default value is set based on the lower bound numbers I am getting in my (and QA) test environment that is prioritizing on *no* memory growth.